onnxruntime/core/optimizer/compute_optimizer/upstream_reshape.cc

```diff
@@ -0,0 +1,271 @@
+           std::make_shared<LayerNormalizationReshapeActor>(), opset_1)},
+      {GetFullQualifiedOpName("MatMul", kOnnxDomain),
+       OpPassThroughConfig<UpStreamReshapeOperatorActorBase>(
+           std::make_shared<MatMulReshapeActor>(), opset_13_9_1)},
+  });
+}
+
+bool UpStreamReshapeGraphTransformer::UpStreamInternal(
+    Graph& graph, std::deque<ReshapeInfo>& queue, Node& current_node, ReshapeInfo& info,
+    const OpPassThroughConfig<UpStreamReshapeOperatorActorBase>& pass_through_config,
+    const logging::Logger& logger) const {
+  const std::string op_type = GetFullQualifiedOpName(current_node.OpType(), current_node.Domain());
+
+  std::vector<int> propagate_input_indices;
+  std::unordered_map<int, std::vector<DimCompareRet>> all_input_cmp_rets;
+  std::function<void(Node & node)> shape_update_func;
+  if (!pass_through_config.actor->PreCheck(current_node, info, logger, propagate_input_indices,
+                                           all_input_cmp_rets, shape_update_func)) {
+    LOG_DEBUG_INFO(logger, "Pre-check failed for " + current_node.Name() + "(" + op_type + ")");
+    return false;
+  }
+
+  if (propagate_input_indices.empty()) {
+    LOG_DEBUG_INFO(logger, "Skip handling current node " + current_node.Name() + "(" + op_type +
+                               ") because the requirement is not met.");
+    return false;
+  }
+
+  for (const int& input_idx : propagate_input_indices) {
+    auto candidate_input_shape = current_node.InputDefs()[input_idx]->Shape();
+    if (candidate_input_shape->dim_size() != 3) {
+      LOG_DEBUG_INFO(logger, "Skip handling current node " + current_node.Name() + "(" + op_type +
+                                 ") because not all candidate inputs have rank = 3.");
+      return false;
+    }
+
+    // For unflatten dims, currently only dim values are supported.
+    for (int k = 2; k < candidate_input_shape->dim_size(); ++k) {
+      if (!candidate_input_shape->dim(k).has_dim_value()) {
+        LOG_DEBUG_INFO(logger, "Skip handling current node " + current_node.Name() +
+                                   ": non-dim-value are not supported yet for unflatten dims.");
+        return false;
+      }
+    }
+  }
+
+  if (!shape_update_func) {
+    LOG_DEBUG_INFO(logger, "Skip handling current node " + current_node.Name() + "(" + op_type +
+                               ") because the shape update function is not set.");
+    return false;
+  }
+
+  // Be noted, once we reach this point after PreCheck, graph modification started, any failure after this should
+  // be reported as ERROR.
+
+  // Reshape infos that are populated into current_node's inputs.
+  std::vector<ReshapeInfo> populated_reshape_infos;
+  populated_reshape_infos.reserve(propagate_input_indices.size());
+  std::unordered_map<int, ReshapeInfo> new_reshape_infos;
+  for (auto input_index : propagate_input_indices) {
+    if (current_node.InputDefs()[input_index]->Shape()->dim_size() != 3) {
```

Comment:
Is this check needed as we process the input_indices in order?

Reply:
thanks! remove it.